MTLS authentication in between gloo and envoy (#2254)
solo-io/gloo: docs/content/security/mtls/_index.md

---

> title: Gloo MTLS mode

Reviewer: I really like the detailed explanation of all the changes that are required to enable this feature, but currently this document feels more like an internal design doc than a user facing guide. I think we could make it clearer even just by renaming the headings. I would suggest we restructure sections in the following way:

1. **Motivation**: brief overview of xDS and why you might want encryption.
2. **Enabling mTLS**: Helm values and a brief explanation of what the effect is (a simple list with the differences with a normal install)
3. **Detailed explanation**: here I would include all the details that we currently have in the guide. I can see a structure like:
    - Secret creation
    - Gloo deployment (the xDS server)
    - The 3 types of xDS clients:
        - gateway-proxy
        - extauth
        - rate limit

Does that make sense?

weight: 25

> description: Gloo MTLS is a way to ensure that communications between Gloo and Envoy is secure. This is useful if your control-plane is in a different environment than your envoy instance.

Reviewer: Since this occurs in several places, can you do a search/replace to turn all instances of `MTLS` to `mTLS`?

---

{{% notice note %}}
This feature was introduced in version 1.3.4 of Gloo and version 1.3.1 of Gloo Enterprise.

> If you are using an earlier version of open source Gloo, please disregard this page.

Reviewer: Nit:
```suggestion
If you are using earlier versions of Gloo, this feature will not be available.
```

{{% /notice %}}

### Architecture

Gloo and Envoy communicate through the [xDS protocol](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#streaming-grpc-subscriptions)
Essentially, Envoy reaches out to Gloo and sets up an open line of communication where it gets the configuration.
As long as this communication is done with TLS, the config secrets will not be in plaintext.

> Essentially, we’re implementing a poor man’s service mesh here.

Author: removed

```

Finally, the 'gloo-mtls-certs' secret is added to the volumes to make it accessible:

```yaml
      volumes:
      - name: gloo-mtls-certs
        secret:
          defaultMode: 420
          secretName: gloo-mtls-certs
```

#### Gloo Settings

The default settings CRD changes such that the gloo.xdsBindAddr will only listen to incoming requests
from localhost.

`k edit settings.gloo.solo.io -n gloo-system default -oyaml`

{{< highlight yaml "hl_lines=2" >}}
  gloo:
    xdsBindAddr: 127.0.0.1:9999
{{< /highlight >}}

The address 127.0.0.1 binds all incoming connections to Gloo to localhost. This ensures that only the envoy
sidecar can connect to the Gloo, but not any other malicious sources.

The Gloo Settings CR gets picked up automatically within ~5 seconds, so there’s no need to restart the Gloo pod.

#### Changes to the xDS clients

##### Gateway-Proxy

The gateway-proxy pod is changed so that Envoy will initialize the connection to Gloo using TLS.

The configmap has the following change:

{{< highlight yaml "hl_lines=2-13" >}}
    clusters:
      - name: gloo.gloo-system.svc.cluster.local:9977
        transport_socket:
          name: envoy.transport_sockets.tls

>           config:

Reviewer: Use `typed_config` here to reflect the actual config
